File: deeplearning4j-ui-parent/deeplearning4j-ui-components/src/main/typescript/org/deeplearning4j/ui/components/chart/Chart.ts
/// <reference path="../../api/Component.ts" />
/// <reference path="../../typedefs/d3.d.ts" />
/// <reference path="../../util/TSUtils.ts" />


import Ordinal = d3.scale.Ordinal;

abstract class Chart extends Component {

    protected style: StyleChart;

    protected title: string;
    protected suppressAxisHorizontal: boolean;
    protected suppressAxisVertical: boolean;
    protected showLegend: boolean;

    protected setXMin: number;
    protected setXMax: number;
    protected setYMin: number;
    protected setYMax: number;

    protected gridVerticalStrokeWidth: number;
    protected gridHorizontalStrokeWidth: number;

    constructor(componentType: ComponentType, jsonStr: string){
        super(componentType);

        var jsonOrig: any = JSON.parse(jsonStr);
        var json: any = jsonOrig[ComponentType[componentType]];

        this.suppressAxisHorizontal = json['suppressAxisHorizontal'];
        this.suppressAxisVertical = json['suppressAxisVertical'];
        this.showLegend = json['showLegend'];

        this.title = json['title'];
        this.setXMin = json['setXMin'];
        this.setXMax = json['setXMax'];
        this.setYMin = json['setYMin'];
        this.setYMax = json['setYMax'];

        this.gridVerticalStrokeWidth = json['gridVerticalStrokeWidth'];
        this.gridHorizontalStrokeWidth = json['gridHorizontalStrokeWidth'];

        if(json['style']) this.style = new StyleChart(json['style']);
    }

    getStyle(): StyleChart {
        return this.style;
    }

    protected static appendTitle(svg: any, title: string, margin: Margin, titleStyle: StyleText): void{
        var text = svg.append("text")
            .text(title)
            .attr("x", (margin.widthExMargins / 2))
            .attr("y", 0 - ((margin.top - 30) / 2))
            .attr("text-anchor", "middle");

        if(titleStyle){
            if(titleStyle.getFont()) text.attr("font-family",titleStyle.getFont);
            if(titleStyle.getFontSize() != null) text.attr("font-size",titleStyle.getFontSize()+"pt");
            if(titleStyle.getUnderline() != null) text.style("text-decoration", "underline");
            if(titleStyle.getColor()) text.style("fill",titleStyle.getColor);
            else text.style("fill",ChartConstants.DEFAULT_TITLE_COLOR);
        } else {
            text.style("text-decoration", "underline");
            text.style("fill",ChartConstants.DEFAULT_TITLE_COLOR);
        }
    }
}

File: deeplearning4j-ui-parent/deeplearning4j-ui-components/src/main/java/org/deeplearning4j/ui/components/chart/Chart.java
package org.deeplearning4j.ui.components.chart;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Data;
import lombok.Getter;
import org.deeplearning4j.ui.api.Component;
import org.deeplearning4j.ui.components.chart.style.StyleChart;

/**
 * Created by Alex on 3/04/2016.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
public abstract class Chart extends Component {

    private String title;
    private Boolean suppressAxisHorizontal;
    private Boolean suppressAxisVertical;
    private boolean showLegend;

    private Double setXMin;
    private Double setXMax;
    private Double setYMin;
    private Double setYMax;

    private Double gridVerticalStrokeWidth;
    private Double gridHorizontalStrokeWidth;

    public Chart(String componentType, Builder builder) {
        super(componentType, builder.getStyle());
        this.title = builder.title;
        this.suppressAxisHorizontal = builder.suppressAxisHorizontal;
        this.suppressAxisVertical = builder.suppressAxisVertical;

        this.setXMin = builder.setXMin;
        this.setXMax = builder.setXMax;
        this.setYMin = builder.setYMin;
        this.setYMax = builder.setYMax;

        this.gridVerticalStrokeWidth = builder.gridVerticalStrokeWidth;
        this.gridHorizontalStrokeWidth = builder.gridHorizontalStrokeWidth;
    }


    @Getter
    public static abstract class Builder<T extends Builder<T>> {

        private String title;
        private StyleChart style;
        private Boolean suppressAxisHorizontal;
        private Boolean suppressAxisVertical;
        private boolean showLegend;

        private Double setXMin;
        private Double setXMax;
        private Double setYMin;
        private Double setYMax;

        private Double gridVerticalStrokeWidth;
        private Double gridHorizontalStrokeWidth;

        public Builder(String title, StyleChart style) {
            this.title = title;
            this.style = style;
        }

        public T suppressAxisHorizontal(Boolean suppressAxisHorizontal) {
            this.suppressAxisHorizontal = suppressAxisHorizontal;
            return (T) this;
        }

        public T suppressAxisVertical(Boolean suppressAxisVertical) {
            this.suppressAxisVertical = suppressAxisVertical;
            return (T) this;
        }

        public T showLegend(boolean showLegend) {
            this.showLegend = showLegend;
            return (T) this;
        }

        public T setXMin(Double xMin) {
            this.setXMin = xMin;
            return (T) this;
        }

        public T setXMax(Double xMax) {
            this.setXMax = xMax;
            return (T) this;
        }

        public T setYMin(Double yMin) {
            this.setYMin = yMin;
            return (T) this;
        }

        public T setYMax(Double yMax) {
            this.setYMax = yMax;
            return (T) this;
        }

        public T setGridWidth(Double gridXStrokeWidth, Double gridYStrokeWidth){
            this.gridVerticalStrokeWidth = gridXStrokeWidth;
            this.gridHorizontalStrokeWidth = gridYStrokeWidth;
            return (T) this;
        }

        public T setGridWidth(Integer gridXStrokeWidth, Integer gridYStrokeWidth){
            return setGridWidth((gridXStrokeWidth != null ? gridXStrokeWidth.doubleValue() : null),
                    (gridYStrokeWidth != null ? gridYStrokeWidth.doubleValue() : null));
        }

    }

}
